mimxrt: Add the machine.uart module.

robert-hh opened the pull request:
This PR adds the machine.UART class.

It uses the LPUARTx devices. Up to 8 UARTS can be used, given that the
Pins are accessible. E.g. 8 on Teensy 4.1, 5 on MIMXRT1020_EVK.
For Teensy 4.0 and 4.1 the UART numbers are as printed on the Pinout
1..n. The MIMXRT10xx-EVK boards have only one UART named, which get the
number 1. All other UART are assigned to different Pins:

MIMXRT1010-EVK:
D0/D1 -> UART 1
D6/D7 -> UART 2
A0/D4 -> UART 3

MIMXRT1020-EVK:
D0/D1   UART 1
D6/D9   UART 2
D10/D12 UART 3
D14/D15 UART 4
A0/A1   UART 5

MIMXRT1050-EVK, MIMXRT1060-EVK, MIMXRT1064-EVK:
D0/D1  UART 1
D7/D6  UART 2
D8/D9  UART 3
A1/A0  UART 4

The UART code has been tested by sending an receiving ~60MByte of data
in 1000 byte blocks at 921600 baud. It did not miss one byte.

Note: The MIMXRT MCUs allow to route the UART signals to various
predefined Pads, and only some of the Pads are routed to boards pins.
And each board pins can have different functions assigned. That ends
up in a pretty complex pad <-> function setup. I decides to code
that very straight, maybe not overly elegant. But it looked more robust
for maintenance.

robert-hh committed:
mimxrt: Add class UART to machine.

Using the LPUARTx devices. Up to 8 UARTS can be used, given that the
Pins ar accessible. E.g. 8 on Teensy 4.1, 5 on MIMXRT1020_EVK.
For Tennsy 4.0 and 4.1 the uart numbers are as printed on the Pinout
1..n. The MIMXRT10xx-EVK boards have only one UART named, which get the
number 1. All other UART are assigned to different Pins:

MIMXRT1010-EVK:
D0/D1 -> UART 1
D6/D7 -> UART 2
A0/D4 -> UART 3

MIMXRT1020-EVK:
D0/D1   UART 1
D6/D9   UART 2
D10/D12 UART 3
D14/D15 UART 4
A0/A1   UART 5

MIMXRT1050-EVK, MIMXRT1060-EVK, MIMXRT1064-EVK:
D0/D1  UART 1
D7/D6  UART 2
D8/D9  UART 3
A1/A0  UART 4

robert-hh committed:
mimxrt uart: Get rid of machine_uart.h.

robert-hh committed:
mimxrt: Fix the code formatting.

robert-hh committed:
mimxrt: Support buffered transmit for UART.

robert-hh commented:
Another implementation question to @dpgeorge. The code implements the typical triple of make_new(), init() and init_helper(). I saw that in the rp2 port you dropped init() and init_helper(). That is reasonable, since the separate init() for UART is a common source of problems. Should I drop the separate init() too?

dpgeorge commented:
> Should I drop the separate init() too?

The intention of init() is to be able to reconfigure individual parameters, like just the baudrate.  If you've implemented it then I think keep it (I still need to review this PR though).

dpgeorge commented on ports/mimxrt/boards/MIMXRT1010_EVK/lpdev_iomux.c:
```diff
@@ -0,0 +1,66 @@
+            IOMUXC_SetPinConfig(IOMUXC_GPIO_10_LPUART1_TXD, 0x10B0u);
+
+            IOMUXC_SetPinMux(IOMUXC_GPIO_09_LPUART1_RXD, 0U);
+            IOMUXC_SetPinConfig(IOMUXC_GPIO_09_LPUART1_RXD, 0x10B0u);
```
Is it possible to reduce all this code duplication (for each board) by providing a table of pins, eg:
```c
pins = {
    {IOMUXC_GPIO_10_LPUART1_TXD, 0U, IOMUXC_GPIO_10_LPUART1_TXD, 0x10B0u, ...}, // LPUART1
    {0, 0, 0, 0, 0, 0, 0, 0}, // LPUART2 not supported
    {....}, // LPUART3
};
```

robert-hh replied:
I had an initial version which was based on tables. That was not efficient and hard to maintain. Your suggestion looks more compact, but it still has to be MCU- and board-specific, because the MCUs have the pins configured at different pins, and the boards have different UARTs exposed to their pins. Also, these IOMUXC_GPIO_10_LPUART1_TXD macros are effectively sets of five values, which have to be expanded when using them.
Besides that, alphafred and me are considering to replace the existing Pin defintion scheme with pins.csv and MIMXRT10xx-af.csv by a method, which parses fsl_iomuxc.h (thats where the above macro is from), replacing the MIMXRT10xx-af.csv files. This file contains all information needed. That would then replace these lpdev_iomuxc.c files. But that is still to come. Alphafred is at the moment busy with implementing SD card support.

dpgeorge replied:
Ok, le'ts keep it as-is for now.

If you can eventually reduce duplicated code among boards, that would be a good goal.

robert-hh replied:
I just started to implement your suggestion to see, if there is a benefit in code size.
The duplication among boards affects Teensy 4.0 and Teensy 4.1, which have a similar board set-up, and MIMXRT1050, 1060 and 1064, which have the same Pin setup on the board. These could re-use the same file.

robert-hh replied:
So, the implementation increased the Teensy 4.0 code by 52 bytes, and decreased the Teensy 4.1 code bye 8 bytes. But the code looks more clean.

dpgeorge replied:
Looks much cleaner!

robert-hh committed:
mimxrt: Use tables for IOMUXC definitions.

The tables are defined in mpconfigboard.h, and the code using it is
in machine_uart.c. That allowed to get rid of the lpdev_iomux.c
files in the board directories.

dpgeorge commented:
Squashed and merged in 689476c576dc4ea5e944f396bd4bf2b95e08cf70 with some minor edits (typos fixed, code style cleaned up a tiny bit).

Thank you!
